go/hyperkit.go

```diff
@@ -0,0 +1,528 @@
+	}
+	if _, err = os.Stat(h.Kernel); os.IsNotExist(err) {
+		return fmt.Errorf("Kernel %s does not exist", h.Kernel)
+	}
+	if _, err = os.Stat(h.Initrd); os.IsNotExist(err) {
+		return fmt.Errorf("initrd %s does not exist", h.Initrd)
+	}
+	if h.DiskImage == "" && h.StateDir == "" && h.DiskSize != 0 {
+		return fmt.Errorf("Can't create disk, because neither DiskImage nor StateDir is set")
+	}
+
+	// Create files
+	if h.StateDir != "" {
+		err = os.MkdirAll(h.StateDir, 0755)
+		if err != nil {
+			return err
+		}
+	}
+	if h.DiskImage == "" && h.DiskSize != 0 {
+		h.DiskImage = filepath.Join(h.StateDir, "disk.img")
+	}
+	if _, err = os.Stat(h.DiskImage); os.IsNotExist(err) {
+		if h.DiskSize != 0 {
+			err = CreateDiskImage(h.DiskImage, h.DiskSize)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if h.UserData != "" {
+		h.ISOImage, err = createUserDataISO(h.StateDir, h.UserData)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Run
+	h.buildArgs(cmdline)
+	err = h.execHyperKit()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Stop the running VM
+func (h *HyperKit) Stop() error {
+	if h.process == nil {
+		return fmt.Errorf("hyperkit process not known")
+	}
+	if !h.IsRunning() {
+		return nil
+	}
+	err := h.process.Kill()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsRunning returns true if the hyperkit process is running.
+func (h *HyperKit) IsRunning() bool {
+	// os.FindProcess on Unix always returns a process object even
+	// if the process does not exists. There does not seem to be
+	// a call to find out if the process is running either, so we
+	// use another package to find out.
+	proc, err := ps.FindProcess(h.Pid)
+	if err != nil {
+		return false
+	}
+	if proc == nil {
+		return false
+	}
+	return true
+}
+
+// Remove deletes all statefiles if present.
+// This also removes the StateDir if empty.
+// If keepDisk is set, the diskimage will not get removed.
+func (h *HyperKit) Remove(keepDisk bool) error {
+	if h.IsRunning() {
+		return fmt.Errorf("Can't remove state as process is running")
+	}
+	if h.StateDir == "" {
+		// If there is not state directory we don't mess with the disk
+		return nil
+	}
+
+	if !keepDisk {
+		return os.RemoveAll(h.StateDir)
+	}
+
+	files, _ := ioutil.ReadDir(h.StateDir)
+	for _, f := range files {
+		fn := filepath.Clean(filepath.Join(h.StateDir, f.Name()))
+		if fn == h.DiskImage {
+			continue
+		}
+		err := os.Remove(fn)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Convert to json string
+func (h *HyperKit) String() string {
+	s, _ := json.Marshal(h)
+	return string(s)
+}
+
+// CreateDiskImage creates a empty file suitable for use as a disk image for a hyperkit VM.
+func CreateDiskImage(location string, sizeMB int) error {
+	f, err := os.Create(location)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	buf := make([]byte, 1048676)
+	for i := 0; i < len(buf); i++ {
+		buf[i] = 0
```

Comment:
https://blog.golang.org/go-slices-usage-and-internals says:
> Arrays do not need to be initialized explicitly; the zero value of an array is a ready-to-use array whose elements are themselves zeroed: 

This is technically a slice, but there must be an underlying array so I think this holds and the init is unnecessary.

http://stackoverflow.com/questions/16797380/how-to-create-a-10mb-file-filled-with-000000-data-in-golang suggests `f.Truncate(size)` might be workable, not sure about on OSX due to lack of sparse files but might save this entire wodge of code.